### PR TITLE
[Code health] Clean up and document observables in MainViewModel

### PR DESCRIPTION
**Pull request by gino-m**
- [x] Make `onSignInStateChange()` fully lazy and `@Cold`
- [x] Add JavaDoc and missing Rx annotations
- [x] Make schedulers specific
- [x] Simplify observers slightly
- [x] Small rename refactor
- [x] Introduce custom operator `switchMapIf()`.

Work towards #699.

### CODE REVIEW

**Comment by gino-m** [comment]
@shobhitagarwal1612 PTAL?

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 1]
```diff
@@ -42,7 +44,12 @@
 @SharedViewModel
 public class MainViewModel extends AbstractViewModel {
 
+  /** The window insets determined by the activity. */
+  @Hot(replays = true)
   private final MutableLiveData<WindowInsetsCompat> windowInsets = new MutableLiveData<>();
+
+  /** The state of sign in progress dialog visibility. */
+  @Hot(replays = true)
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java:54]: Same as above.
Annotation already added to the public accessible method. Let's keep it at only one place.

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 2]
```diff
@@ -42,7 +44,12 @@
 @SharedViewModel
 public class MainViewModel extends AbstractViewModel {
 
+  /** The window insets determined by the activity. */
+  @Hot(replays = true)
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java:50]: Annotation already added to the public accessible method. Let's keep it at only one place.

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 3]
```diff
@@ -67,11 +75,16 @@ public MainViewModel(
 
     // TODO: Move to background service.
     disposeOnClear(
-        projectRepository.getActiveProject().switchMapCompletable(this::syncFeatures).subscribe());
+        projectRepository
+            .getActiveProject()
+            .observeOn(schedulers.io())
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java:82]: Do we need to pass the `subscribeOn` scheduler as well?

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 4]
```diff
@@ -67,11 +75,16 @@ public MainViewModel(
...
+            .switchMapCompletable(this::syncFeatures)
+            .subscribe());
 
     disposeOnClear(
         authenticationManager
             .getSignInState()
+            .observeOn(schedulers.ui())
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java:90]: `.observeOn(schedulers.ui())` 
Why is this change needed?

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 5]
```diff
@@ -95,8 +109,19 @@ void onApplyWindowInsets(WindowInsetsCompat insets) {
     windowInsets.setValue(insets);
   }
 
+  @Cold
   private Completable onSignInStateChange(SignInState signInState) {
     Timber.d("Auth status change: %s", signInState.state());
+    Optional<User> user = signInState.getUser();
+
+    if (user.isPresent()) {
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java]: We should also check for state == `SIGNED_IN`

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 6]
```diff
@@ -138,9 +159,11 @@ void onSignedOut() {
     navigator.navigate(SignInFragmentDirections.showSignInScreen());
   }
 
-  void onSignedIn() {
-    hideProgressDialog();
-    navigator.navigate(HomeScreenFragmentDirections.showHomeScreen());
+  private Completable onSignedIn(User user) {
+    return userRepository
+        .saveUser(user)
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java]: Can we move the user creation under if block `onSignInStateChange` and call `onSignedIn` from there?

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java [hunk 7]
```diff
@@ -111,13 +112,8 @@ void onApplyWindowInsets(WindowInsetsCompat insets) {
...
   @Cold
-  private Observable<SignInState> saveUserOnSignIn(SignInState signInState) {
-    Optional<User> user = signInState.getUser();
-    if (user.isPresent()) {
-      return userRepository.saveUser(user.get()).andThen(Observable.just(signInState));
-    } else {
-      return Observable.just(signInState);
-    }
+  private Completable saveUser(SignInState signInState) {
+    return userRepository.saveUser(signInState.getUser().get());
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/MainViewModel.java:116]: `.get()` -> `.orElseThrow()` ?

File: gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.java [hunk 8]
```diff
@@ -133,7 +133,7 @@ public ProjectRepository(
     return projectLoadingState.map(Loadable::value);
   }
 
-  public void onSelectProject(String projectId) {
+  public void activateProject(String projectId) {
```
Comment (shobhitagarwal1612) [inline gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.java:136]: How about adding a `@NonNull` to the parameter and removing `Optional` below? 

I think all the usages of this method are already checking for non-null project ids. So we can simplify it here.

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]
Sorry for the delay. Some comments for your consideration.

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java (hunk 1 above)
Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java:54]: I think we should annotation both fields and accessors. In many case the class' impl will interact with the field directly (not via the public accessor), in which case it will be important to understand the observable's semantics.

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java (hunk 2 above)
Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java:50]: See previous comment.

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java (hunk 3 above)
Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java:82]: The WIP best practices doc prescribes that we specify `subscribeOn()` in the method that defines the background operation (in this case `getActiveProject()`).

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java (hunk 4 above)
Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java:90]: Our best practices guidelines recommend specifying `observeOn()`  downstream close to `subscribe()`. But I think in this case it's irrelevant to place it here since we only do work inside the `switchMap` operation, which has its own subscription and threads. I'll remove it to avoid confusion.

Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java:90]: It's needed to ensure we're actually using the UI thread to process values handled by `subscribe()`.

File: gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.java (hunk 8 above)
Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.java:136]: `@NonNull` is the default in our current setup, so we don't need to specify it.

The `Optional` isn't being used to check for nulls here.. not sure I follow?

**Review by gino-m — COMMENTED** [review]
Thanks @shobhitagarwal1612 for reviewing. Sent replies and follow up questions. PTAL.

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java (hunk 7 above)
Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/MainViewModel.java]: Good point.. we shouldn't use `get()` unless we explicitly check whether the Optional is empty. The problem here is that the check would need to be done in the caller, not here together with the call to `get()`.

What if we consolidate this logic into our custom operators? Then the check and accessor are called in the same place, and the code is a bit more readable. Added a commit to demonstrate. Wdyt?

File: gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.java (hunk 8 above)
Comment (shobhitagarwal1612) [reply gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.java:136]: `selectProjectEvent` wraps the `projectId` as an Optional. As you already mentioned, the string can't be null. So I was wondering if we can remove the optional now. Line 138

**Review by shobhitagarwal1612 — APPROVED** [review]
